### DISCUSSION TITLE
How to use different mocks for each test with vi.mock?

### DISCUSSION THREAD

**Discussion by marcadrian-it** [vitest-dev/vitest#3589] [Q&A] [answered]
Hello, I’m having trouble using different mocks for each test in my test suite. 
Here’s my code:
```javascript
import { render, screen } from '@testing-library/react';
import { vi } from 'vitest';
import HomePage from '../app/page';

describe('Homepage', () => {
  test('Home page renders correctly', async () => {
    render(await HomePage());
    expect(screen.getByText('get started')).toBeInTheDocument();
  });

  test('Conditional rendering when authenticated', async () => {
    vi.mock('@clerk/nextjs', () => {
      return {
        auth: () => new Promise((resolve) => resolve({ userId: 'fefsfehrf' })),
        ClerkProvider: ({ children }) => <div>{children}</div>,
        useUser: () => ({
          isSignedIn: true,
          user: {
            id: 'user_8JkL2mP0zX6d8JkL2mP0zX6dJ',
            fullName: 'Thrall Durotan',
          },
        }),
      };
    });

    const { debug } = render(await HomePage());
    debug();
    const link = screen.getByRole('link');
    expect(link).toHaveAttribute('href', '/journal');
  });

  test('Conditional rendering when not authenticated', async () => {
    vi.mock('@clerk/nextjs', () => {
      return {
        auth: () => new Promise((resolve) => resolve({ userId: null })),
        ClerkProvider: ({ children }) => <div>{children}</div>,
        useUser: () => ({
          isSignedIn: false,
          user: null,
        }),
      };
    });

    const { debug } = render(await HomePage());
    debug();
    const link = screen.getByRole('link');
    expect(link).toHaveAttribute('href', '/new-user');
  });
});
 
```
I’m calling vi.mock twice, once in the second test and once in the third test. However, it seems that the second call to vi.mock is overwriting the first mock as when I console log userId both tests use it mocked as null, which means both tests are using the mock defined in the second call.

I’ve tried calling vi.resetAllMocks in an afterEach block and setting mockReset: true in my config, but it's not really problem of a mock reset rather some overwrite problem.

Can you help me understand how to use different mocks for each test with vi.mock? Is there something I’m missing or doing wrong?

Thank you!

**Comment by sheremet-va** [comment]
https://vitest.dev/api/vi.html#vi-mock

> The call to `vi.mock` is hoisted, so it doesn't matter where you call it. It will always be executed before all imports.

**Reply by marcadrian-it** [reply]
Thank you for explaining that the call to vi.mock is hoisted.
However, I’m still not quite sure how to make it so that vi.mock uses different parameters for each test. I’m entirely new to mocking, so any guidance or examples would be greatly appreciated. Thank you!

**Reply by sheremet-va** [reply, marked as answer]
You can pass down variables with `vi.hoisted` and change the implementation before running a test:

```ts
import { namedExport } from './path/to/module.js'

const mocks = vi.hoisted(() => {
  return {
    namedExport: vi.fn(),
  }
})

vi.mock('./path/to/module.js', () => {
  return {
    namedExport: mocks.namedExport,
  }
})

it('test 1', () => {
	mocks.namedExports.mockReturned(true)
	// some test
})

it('test 2', () => {
	mocks.namedExports.mockReturned(false)
	// some test
})
```

**Reply by sinaza** [reply]
I've had the same issue trying to mock `@remix-run/react`'s `useParams` and 👆 came to the rescue. In my case, I also had to call `mockImplementation` instead of `mockReturned` e.g.

```javascript
  mocks.useParams.mockImplementation(() => ({
    "*": "",
  }));
```

**Reply by tnt-glenn-mccomb** [reply]
Thanks for including this additional context @sinaza, I found it very useful when trying to mock `react-router-dom`'s `useLocation` hook since `.mockReturned` didn't work.

**Reply by Sti2nd** [reply]
Remember to restore the mocks at the end of the test!

```js
import { namedExport } from './path/to/module.js'

const mocks = vi.hoisted(() => {
  return {
    namedExport: vi.fn(),
  }
})

vi.mock('./path/to/module.js', () => {
  return {
    namedExport: mocks.namedExport,
  }
})

it('test 2', () => {
  mocks.namedExports.mockReturned(false)
  // some test
  mocks.namedExport.mockRestore() // <-- Important
})

it('test 3', () => {
  // some test here will use namedExport: vi.fn() because we remembered to restore the mocks
}
```

**Reply by abdelrahman543873** [reply]
herooooo thanks so much working

**Reply by seahindeniz** [reply]
This suggestion doesn't work with in-source testing
```bash
Error: There are some problems in resolving the mocks API.
You may encounter this issue when importing the mocks API from another module other than 'vitest'.
To fix this issue you can either:
- import the mocks API directly from 'vitest'
- enable the 'globals' options
```

**Reply by Hugos68** [reply]
Great suggestion!

**Comment by rocifier** [comment]
> You can pass down variables with `vi.hoisted` and change the implementation before running a test:
> 
> ```ts
> import { namedExport } from './path/to/module.js'
> 
> const mocks = vi.hoisted(() => {
>   return {
>     namedExport: vi.fn(),
>   }
> })
> 
> vi.mock('./path/to/module.js', () => {
>   return {
>     namedExport: mocks.namedExport,
>   }
> })
> 
> it('test 1', () => {
> 	mocks.namedExports.mockReturned(true)
> 	// some test
> })
> 
> it('test 2', () => {
> 	mocks.namedExports.mockReturned(false)
> 	// some test
> })
> ```

What if you don't have a named export, eg. a vue component: `import MyComponent from '@/components/MyComponent.vue'`

**Reply by sheremet-va** [reply]
It is called default export and it's also mentioned in documentation. Just use the name "default"

**Comment by posti85** [comment]
Instead of the accepted answer, wouldn't be easier like this? 🤔

```js
import { namedExport } from './path/to/module.js'

vi.mock('./path/to/module.js', () => ({
  namedExport: vi.fn()
}))

it('test 1', () => {
  namedExport.mockImplementation(true)
  // some test
});

it('test 2', () => {
  namedExport.mockImplementation(false)
  // some test
});
```

in TypeScript project, use:
```ts
vi.mocked(namedExport).mockImplementation(...)
```

**Comment by chiptus** [comment]
what if my module exports a const:

util.ts
```
export const isRight = ENV_VAR === "true" ? true : false
```

module
```
import {isRight} from 'util.ts'

export function checkWhatIsRight() {
   return isRight ? 'right' : 'left'
}
```

module.test.ts
```
test("should be 'right' when isRight", () => {
   expect (checkWhatIsRight()).toBe('right')
})

test("should be 'left' when not isRight", () => {
   expect (checkWhatIsRight()).toBe('left')
})
```

**Reply by chiptus** [reply]
I tried:

```
const mocks = vi.hoisted(() => ({ isRight:false }))

vi.mock('util.ts', () => ({
   isRight: mocks.isRight
}))
```

**Reply by muraha** [reply]
But it's const. You are not suppose to change it. Or if we assume it's `let` and some other function changes it, just mock those fn and run mocked fn before the actual test.

**Reply by chiptus** [reply]
I've updated the code, to better represent what I'm doing.

**Reply by chiptus** [reply]
and I don't have a function that gets or changes it. but I guess that's a solution, to use a getter

**Reply by muraha** [reply]
then `vi.stubEnv()` is your way to go: https://vitest.dev/api/vi.html#vi-stubenv
